### PR TITLE
API: add return_inverse to pd.unique

### PR DESCRIPTION
**Pull request by h-vetinari**
- [x] splits off first chunk of #24108; progress towards #4087 / #21357 / #21720 / #22824
- [x] tests expanded / parametrized / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This is the first part I'm splitting off of #24108, but now with full test coverage. For the moment, I've added `return_inverse` to `pd.unique` and to `Categorical.unique`, but it's not trivial because of inconsistencies like the following:
```
>>> import pandas as pd
>>> idx = pd.Index([0, 1, 1, 0])
>>> pd.unique(idx)
array([0, 1], dtype=int64)
>>>
>>> # So pd.unique(Index) yields an array, except if the Index is categorical...?
>>> idx = idx.astype('category')
>>> pd.unique(idx)
CategoricalIndex([0, 1], categories=[0, 1], ordered=False, dtype='category')
```

I'd be open to further split off the change for `Categorical.unique`, or just return `NotImplemented` for all `ExtensionArray` types. As mentioned in #24108 already, I believe that the possibility for `return_inverse` (or maybe even kwargs in general??) is something that should be added to the EA interface. @TomAugspurger @jreback @jbrockmendel

### CODE REVIEW

**Comment by h-vetinari** [comment]
Thinks changed around a bit here with `DatetimeArray` etc., but this commit should work.

As mentioned in the OP, this splits off first chunk of #24108 and makes some progress towards #4087 / #21357 / #21720 / #22824. I'm sure there'll still be lots of discussion, but having an implementation is a good start (even though there's not much happening - the cython backend is already there since a few months).

The diff in `test_algos.py` is completely busted, but I painstakingly created some nicely modular commits to step through the changes in a sane fashion.

**Comment by h-vetinari** [comment]
@jreback @jorisvandenbossche @TomAugspurger 
Is it possible to give this PR some initial review? It's been lying around for ~3 months...

**Comment by TomAugspurger** [comment]
I won't have time in the near-term.

**Comment by jreback** [comment]
so again why should we add a method to do this, when we already have one?

```
In [16]: idx = pd.Index([0, 1, 1, 0])                                                                                                                                                                                                   

In [17]: pd.factorize(idx)                                                                                                                                                                                                              
Out[17]: (array([0, 1, 1, 0]), Int64Index([0, 1], dtype='int64'))

In [18]: idx.take(pd.factorize(idx)[0])                                                                                                                                                                                                 
Out[18]: Int64Index([0, 1, 1, 0], dtype='int64')
```

true numpy calls this ``return_inverse`` in ``.unique()`` but they also don't have ``.factorize()``. I really don't like having multiple ways of doing things. I am not convinced this is actually that useful. Can you show the usecase that is not *possible* or (claimed) not performant that this cannot be done now?

**Review by jreback — CHANGES_REQUESTED** [review]
comments

**Comment by jreback** [comment]
closing as stale

**Comment by h-vetinari** [comment]
@jreback, this is not stale, but I do see that I overlooked to answer your comment a bit further up.

I've made the case in #22824 several times, but among other things, `factorize` and `unique` do *different* things, namely in the treatment of missing values (and there are several cases where that difference can be crucial).

Then there's the fact that most people who are not knee-deep in stats or R won't know `factorize`, but `unique` is incredibly intuitive and matches with what `numpy` offers.

Please reopen this and lets have this discussion (bearing in mind that the actual goal would be #24108; this PR is just a stepping stone).

**Comment by h-vetinari** [comment]
Thanks for reopening! Will try to merge master soon.

File: doc/source/whatsnew/v1.0.0.rst [hunk 1]
```diff
@@ -94,6 +94,25 @@ of the Series or columns of a DataFrame will also have string dtype.
 We recommend explicitly using the ``string`` data type when working with strings.
 See :ref:`text.types` for more.
 
+
+.. _whatsnew_1000.enhancements.unique:
```
Comment (h-vetinari) [inline doc/source/whatsnew/v1.0.0.rst]: Note, this was chopped off of #24108 and the section is intended to be bigger, compare [here](https://github.com/pandas-dev/pandas/pull/24108/files)

File: pandas/tests/test_algos.py [hunk 2]
```diff
@@ -355,19 +371,23 @@ def test_factorize_na_sentinel(self, sort, na_sentinel, data, uniques):
         else:
             tm.assert_extension_array_equal(uniques, expected_uniques)
 
-
 class TestUnique:
```
Comment (h-vetinari) [inline pandas/tests/test_algos.py]: Note that no tests are removed here (even though the diff is large). I sometimes joined tests and parametrized them with fixtures. In fact, there should be many more tests now...

File: pandas/tests/test_algos.py [hunk 3]
```diff
@@ -450,102 +472,87 @@ def test_nan_in_object_array(self):
...
-        )
-        expected = DatetimeIndex(
-            ["2016-01-01 00:00:00"], dtype="datetime64[ns, US/Eastern]", freq=None
-        )
-        tm.assert_index_equal(result, expected)
+        result = method(obj)
+        assert_series_or_index_or_array_or_categorical_equal(result, expected)
+
+        # TODO: add support for return_inverse to DatetimeArray/DatetimeIndex,
+        # as well as [[Series/Index].unique
```
Comment (h-vetinari) [inline pandas/tests/test_algos.py]: I left this here as an indication where things should be heading.

**Review by h-vetinari — COMMENTED** [review]
A few comments on the diff

**Comment by h-vetinari** [comment]
@TomAugspurger @jorisvandenbossche @jreback 
This is updated and green. Since the diff in the tests basically unreadable, I'll recall the [following](https://github.com/pandas-dev/pandas/pull/24119#issuecomment-459917557):

> The diff in test_algos.py is completely busted, but I painstakingly created some nicely modular commits to step through the changes in a sane fashion.

**Comment by jreback** [comment]
I am highly u likely to change my -1 on this

i view this as duplicative and confusing api

if you want to document in an example great

**Comment by h-vetinari** [comment]
> @jreback: i view this as duplicative and confusing api

`.unique` is central API, and extremely well-established (on top of being very intuitive; plus: the method exists in numpy with the inverse). You yourself [suggested](https://github.com/pandas-dev/pandas/issues/4087) to add a way of doing `return_inverse=True` for `unique`.

You had also already agreed to the utility of having an inverse [here](https://github.com/pandas-dev/pandas/issues/21357#issuecomment-399043089) (at the time you suggested to add the inverse to `duplicated`, which I did in #21645, only to be redirected to add the inverse to `.unique` where it fits more properly).

And as I won't tire of repeating: `unique` != `factorize`. The methods have different goals and differ in several key points.

I feel the conversation keeps going in circles - maybe this could be a nice example case for writing a fully fledged [enhancement proposal](https://github.com/pandas-dev/pandas/issues/28568)?

**Comment by jreback** [comment]
closing. I don't think anyone has the bandwith to work with you on this.

**Pull request closed without merging**